src/backend/distributed/shardsplit/shardsplit_logical_replication.c

```diff
@@ -0,0 +1,750 @@
+	PGresult *result = NULL;
+	int response = ExecuteOptionalRemoteCommand(sourceConnection,
+												dropReplicationSlotCommand->data,
+												&result);
+
+	PQclear(result);
+	ForgetResults(sourceConnection);
+
+	/*
+	 * Note: Temporary slot are only live during the session's lifetime causing them to be dropped when the session ends.
+	 */
+	StringInfo createReplicationSlotCommand = makeStringInfo();
+
+	appendStringInfo(createReplicationSlotCommand,
+					 "CREATE_REPLICATION_SLOT %s LOGICAL citus EXPORT_SNAPSHOT;",
+					 ShardSplitTemplateReplicationSlotName(
+						 shardIntervalToSplit->shardId));
+
+	response = ExecuteOptionalRemoteCommand(sourceConnection,
+											createReplicationSlotCommand->data, &result);
+
+	if (response != RESPONSE_OKAY || !IsResponseOK(result) || PQntuples(result) != 1)
+	{
+		ReportResultError(sourceConnection, result, ERROR);
+	}
+
+	/*'snapshot_name' is second column where index starts from zero.
+	 * We're using the pstrdup to copy the data into the current memory context */
+	char *snapShotName = pstrdup(PQgetvalue(result, 0, 2 /* columIndex */));
+	return snapShotName;
+}
+
+
+/*
+ * ShardSplitTemplateReplicationSlotName returns name of template replication slot
+ * following the shard split naming scheme.
+ */
+char *
+ShardSplitTemplateReplicationSlotName(uint64 shardId)
+{
+	return psprintf("%s%lu", SHARD_SPLIT_TEMPLATE_REPLICATION_SLOT_PREFIX, shardId);
+}
+
+
+/*
+ * CreateReplicationSlots creates copies of template replication slot
+ * on the source node.
+ *
+ * sourceNodeConnection - Source node connection.
+ *
+ * templateSlotName - Template replication slot name whose copies have to be created.
+ *                    This slot holds a LSN from which the logical replication
+ *                    begins.
+ *
+ * shardSplitSubscriberMetadataList - List of 'ShardSplitSubscriberMetadata. '
+ *
+ * 'ShardSplitSubscriberMetadata' contains replication slot name that is used
+ *  to create copies of template replication slot on source node. These slot names are returned by
+ * 'worker_split_shard_replication_setup' UDF and each slot is responsible for a specific
+ * split range. We try multiple attemtps to clean up these replicaton slot copies in the
+ * below order to be on safer side.
+ * 1. Clean up before starting shard split workflow.
+ * 2. Implicitly dropping slots while dropping subscriptions.
+ * 3. Explicitly dropping slots which would have skipped over from 2.
+ */
+void
+CreateReplicationSlots(MultiConnection *sourceNodeConnection, char *templateSlotName,
+					   List *shardSplitSubscriberMetadataList)
+{
+	ShardSplitSubscriberMetadata *subscriberMetadata = NULL;
+	foreach_ptr(subscriberMetadata, shardSplitSubscriberMetadataList)
+	{
+		char *slotName = subscriberMetadata->slotInfo->slotName;
+
+		StringInfo createReplicationSlotCommand = makeStringInfo();
+
+		appendStringInfo(createReplicationSlotCommand,
+						 "SELECT * FROM  pg_copy_logical_replication_slot ('%s','%s')",
```

Comment:
would be good to:
- qualify function names using pg_catalog everywhere (we guard against search_path overloading, but oss users might not)
- not have quotes in format strings, but instead use quote_literal_cstr, since quotes in the format string often points to an injection bug and it takes time to understand whether it is buggy or not